include/rmm/mr/device/pool_memory_resource.hpp

```diff
@@ -95,6 +95,37 @@ class pool_memory_resource final
 
     initialize_pool(initial_pool_size, maximum_pool_size);
   }
+  
+  /**
+   * @brief Construct a `pool_memory_resource` and allocate the initial device memory pool using
+   * `upstream_mr`.
+   *
+   * @throws rmm::logic_error if `upstream_mr == nullptr`
+   * @throws rmm::logic_error if `initial_pool_size` is neither the default nor aligned to a
+   * multiple of pool_memory_resource::allocation_alignment bytes.
+   * @throws rmm::logic_error if `maximum_pool_size` is neither the default nor aligned to a
+   * multiple of pool_memory_resource::allocation_alignment bytes.
+   *
+   * @param upstream_mr The memory_resource from which to allocate blocks for the pool.
+   * @param initial_pool_size Minimum size, in bytes, of the initial pool. Defaults to half of the
+   * available memory on the current device.
+   * @param maximum_pool_size Maximum size, in bytes, that the pool can grow to. Defaults to all
+   * of the available memory on the current device.
+   */
+  explicit pool_memory_resource(Upstream& upstream_mr,
```

Comment:
Should we constrain the ctor or the class template itself for `resource<Upstream>`?

Reply:
We can at least constrain this ctor since it will only be used with new types, right?